File: 01-Programming-Languages/JavaScript/Algorithms/String/palindrome.js
/*
Problem: Palindrome Check
Description: Check if a given string is a palindrome (reads the same forwards and backwards).
Ignore case and non-alphanumeric characters.

Approach:
- Use two pointers from start and end, skip non-alphanumeric chars

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: "A man, a plan, a canal: Panama"
Output: true
*/

function isPalindrome(s) {
  let left = 0, right = s.length - 1;
  while (left < right) {
    while (left < right && !isAlphanumeric(s[left])) left++;
    while (left < right && !isAlphanumeric(s[right])) right--;
    if (s[left].toLowerCase() !== s[right].toLowerCase()) return false;
    left++;
    right--;
  }
  return true;
}

function isAlphanumeric(ch) {
  return /[a-zA-Z0-9]/.test(ch);
}

console.log('isPalindrome("A man, a plan, a canal: Panama"):', isPalindrome('A man, a plan, a canal: Panama'));
console.log('isPalindrome("race a car"):', isPalindrome('race a car'));
console.log('isPalindrome("hello"):', isPalindrome('hello'));

File: 01-Programming-Languages/Swift/Algorithms/String/palindrome.swift
/*
Problem: Palindrome Check
Description: Check if a given string is a palindrome (reads the same
           forwards and backwards). Ignore non-alphanumeric characters
           and case.

Approach:
- Use two pointers (left, right) moving inward
- Skip non-alphanumeric characters
- Compare characters case-insensitively

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: "A man, a plan, a canal: Panama"
Output: true
*/

func isPalindrome(_ s: String) -> Bool {
    let chars = Array(s.lowercased().filter { $0.isLetter || $0.isNumber })
    var left = 0, right = chars.count - 1
    while left < right {
        if chars[left] != chars[right] { return false }
        left += 1
        right -= 1
    }
    return true
}

print("Palindrome 'A man, a plan, a canal: Panama': \(isPalindrome("A man, a plan, a canal: Panama"))")
print("Palindrome 'race a car': \(isPalindrome("race a car"))")
